File: TestLeetCode/Tree/LCABst.cs
public class LowestCommonAncestorBstSolution {
    public TreeNode LowestCommonAncestorIter(TreeNode root, TreeNode p, TreeNode q) {
        while (true) {
            if (root.val < p.val && root.val < q.val) {
                root = root.right;
            } else if (root.val > p.val && root.val > q.val) {
                root = root.left;
            } else {
                return root;
            }
        }
    }

    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // Traverse Right child
        if (p.val > root.val && q.val > root.val) {
            return LowestCommonAncestor(root.right, p, q);
        }
        
        // Traverse Left Child
        if (p.val < root.val && q.val < root.val) {
            return LowestCommonAncestor(root.left, p, q);
        }
        
        return root;
    }
}

File: python/tree/lca_bst.py
from python.common import TreeNode


class LowestCommonAncestorBST:
    def lowest_common_ancestor(self, root: TreeNode, left: TreeNode, right: TreeNode) -> TreeNode:
        if left.val < root.val and right.val < root.val: return self.lowest_common_ancestor(root.left, left, right)
        if left.val > root.val and right.val > root.val: return self.lowest_common_ancestor(root.right, left, right)
        return root

    def lowest_common_ancestor_iterative(self, root: TreeNode, left: TreeNode, right: TreeNode) -> TreeNode:
        while (left.val - root.val) * (right.val - root.val) > 0: root = root.left if left.val < root.val else root.right
        return root
